File: library/src/androidTest/java/com/james602152002/floatinglabelspinner/popupwindow/SpinnerPopupWindowTest.java
package com.james602152002.floatinglabelspinner.popupwindow;

import android.content.Context;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnitRunner;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.james602152002.floatinglabelspinner.FloatingLabelSpinner;
import com.james602152002.floatinglabelspinner.adapter.HintAdapter;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.lang.reflect.Field;

/**
 * Created by shiki60215 on 18-1-10.
 */
public class SpinnerPopupWindowTest extends AndroidJUnitRunner {

    private SpinnerPopupWindow popupWindow;
    private FloatingLabelSpinner spinner;

    @Before
    public void setUp() throws Exception {
        final Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        spinner = new FloatingLabelSpinner(context);
        spinner.setDropDownHintView(new View(context));
        spinner.setHint("hint");
        popupWindow = new SpinnerPopupWindow(context);
        spinner.setAdapter(new BaseAdapter() {
            @Override
            public int getCount() {
                return 0;
            }

            @Override
            public Object getItem(int position) {
                return null;
            }

            @Override
            public long getItemId(int position) {
                return 0;
            }

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                return null;
            }
        });
        popupWindow.setAdapter(spinner, (HintAdapter) spinner.getAdapter(), (short) 0, spinner.getOnItemSelectedListener());
    }

    @Test
    public void testOnItemClick() throws NoSuchFieldException, IllegalAccessException {
        popupWindow.onItemClick(null, null, 0, 0);
        Field field = SpinnerPopupWindow.class.getDeclaredField("listener");
        field.setAccessible(true);
        field.set(popupWindow, null);
        spinner.setRecursiveMode(true);
        popupWindow.onItemClick(null, null, 0, 0);
    }

    @After
    public void tearDown() throws Exception {
        popupWindow = null;
        spinner = null;
    }
}

File: library/src/androidTest/java/com/james602152002/floatinglabelspinner/popupwindow/SpinnerPopupWindowTest.kt
package com.james602152002.floatinglabelspinner.popupwindow

import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnitRunner
import com.james602152002.floatinglabelspinner.FloatingLabelSpinner
import com.james602152002.floatinglabelspinner.adapter.HintAdapter
import org.junit.After
import org.junit.Before
import org.junit.Test

/**
 * Created by shiki60215 on 18-1-10.
 */
class SpinnerPopupWindowTest : AndroidJUnitRunner() {
    private var popupWindow: SpinnerPopupWindow? = null
    private var spinner: FloatingLabelSpinner? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        spinner = FloatingLabelSpinner(context)
        spinner!!.dropDownHintView = View(context)
        spinner!!.setHint("hint")
        popupWindow = SpinnerPopupWindow(context)
        spinner!!.setAdapter(object : BaseAdapter() {
            override fun getCount(): Int {
                return 0
            }

            override fun getItem(position: Int): Any? {
                return null
            }

            override fun getItemId(position: Int): Long {
                return 0
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
                return null
            }
        })
        popupWindow!!.setAdapter(
            spinner!!,
            spinner!!.adapter as HintAdapter,
            0,
            spinner!!.onItemSelectedListener
        )
    }

    @Test
    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    fun testOnItemClick() {
        popupWindow!!.onItemClick(null, null, 0, 0)
        val field = SpinnerPopupWindow::class.java.getDeclaredField("listener")
        field.isAccessible = true
        field[popupWindow] = null
        spinner!!.recursiveMode = true
        popupWindow!!.onItemClick(null, null, 0, 0)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        popupWindow = null
        spinner = null
    }
}